fix(filetype): make .tex filetype detection match Vim

gi1242 opened the pull request:
Fix #21594

Synced detection logic to match Vim9 for files with a pre-compilied header.

smjonas commented on runtime/lua/vim/filetype/detect.lua:
```diff
@@ -1256,17 +1256,16 @@ end
-    elseif format == 'plaintex' then
-      return 'plaintex'
-    end
+  local matched, idx, mstr = getlines(bufnr, 1):find('^%%&%s*(%a+)')
```
Thanks for the PR! I would prefer to only change this line to `local _, _, format = getlines(bufnr, 1):find('^%%&%s*(%a+)')` though. The other changes such as directly returning the filetype were made deliberately to simplify the logic compared to Vim's implementation.

gi1242 replied:
The other changes you had unfortunately ended up changing the logic! If format didn't match in the first few lines, the fall back to `latex` is done by Vim, but not by the neovim you had before.

Just try whatever code you end up using on a file that has `%&preamble` in the first file. It should be detected as LaTeX (Vim does it).

clason replied:
That's fine, but the tests (for more typical situations) -- which are taken verbatim from vim! -- must pass as well.

gi1242 replied:
fixed now

zeertzjq commented:
test failure

clason commented:
Specifically, failure of Vim's filetype detection test, which means that the detection logic changed.

gi1242 commented:
Sorry, I'm not familiar with the build process, so can't see what test failed. Can I suggest adding a file `precompiled.tex` with contents:

    %& preamble
    \begin{document}
      Stuff
    \end{document}
    
Vim opens it as LaTeX (as it should), and neovim did not. This patch fixes it.

zeertzjq commented:
The test failure is for `tex/context/.*/.*.tex`

gi1242 commented:
I got your build system running on my machine and tried making tests. Can I get some help. I tried various combinations of `TEST_FILTER="..." make functionaltest` and couldn't narrow it down to the test that failed. Just running `make functionaltest` says

    CMake Error at /home/gautam/tmp/neovim/cmake/RunTests.cmake:107 (message):
      functional tests failed with error: 1


    FAILED: CMakeFiles/functionaltest /home/gautam/tmp/neovim/build/CMakeFiles/functionaltest
    
I couldn't find log files, or instructions on how to find the test that failed...

Thanks

zeertzjq commented:
You don't need to run functionaltests. You need to run `make oldtests TEST_FILE=test_filetype.vim`

gi1242 commented:
Thanks. I had a typo (wrote `contex` instead of `context`. My bad. Tests pass now (on my system)

zeertzjq commented on runtime/lua/vim/filetype/detect.lua:
```diff
@@ -1279,22 +1278,21 @@ function M.tex(path, bufnr)
+  end -- if matched
+
+  -- Translation from formats to file types.  TODO:  add AMSTeX, RevTex, others?
+  if format == 'plaintex' or format == 'context' then
```
This changes the behavior of `g:tex_flavor = 'plain'`. With `g:tex_flavor = 'plain'` filetype should be set `plaintex`, but with this PR it is set to `tex`.

gi1242 replied:
You're right, my apologies. I just pushed a fix.

smjonas commented:
I have created a PR on the Vim repo that adds more test cases: https://github.com/vim/vim/pull/11765. This should contain all things mentioned in this thread.

clason committed:
vim-patch:9.0.1120: tex filetype detection not sufficiently tested

Problem:    Tex filetype detection not sufficiently tested.
Solution:   Add more test cases for "tex" detection. (Jonas Strittmatter,
            closes vim/vim#11765)

https://github.com/vim/vim/commit/c55e8f2c6f5cafe11494df3e2d28ff3d03b92c71

Co-authored-by: smjonas <jonas.strittmatter@gmx.de>

clason commented:
@gi1242 could you cherry-pick https://github.com/neovim/neovim/pull/21601/commits/cc70649f6882fc6b12badcaeffd5cc93eeafe6eb (and squash your other commits?)

gi1242 commented:
Done... (I think. I had to force push, and hopefully it updates here. If not it's f3e23ee2e3 in gi1242/neovim). Thanks for helping me out with your build system

clason commented:
That squash worked -- but was too eager, the vim-patch commit should have been left separate!

Commit:
fix(filetype): correctly detect tex files

Fixes Issue #21594.

gi1242 commented:
10th time a charm?

clason commented:
Sorry, still not quite: you need to cherry-pick the commit from _my_ PR, otherwise it doesn't count as a ported Vim patch.

gi1242 commented:
Ooof. Fixed...

clason commented:
Test failure:
```
From test_filetype.vim:
[4818](https://github.com/neovim/neovim/actions/runs/3812943263/jobs/6486418411#step:14:4819)
Found errors in Test_tex_file():
[4819](https://github.com/neovim/neovim/actions/runs/3812943263/jobs/6486418411#step:14:4820)
command line..script /home/runner/work/neovim/neovim/src/nvim/testdir/runtest.vim[437]..function RunTheTest[44]..Test_tex_file line 10: Expected 'tex' but got 'plaintex'
[4820](https://github.com/neovim/neovim/actions/runs/3812943263/jobs/6486418411#step:14:4821)
command line..script /home/runner/work/neovim/neovim/src/nvim/testdir/runtest.vim[437]..function RunTheTest[44]..Test_tex_file line 32: Expected 'context' but got 'tex'
[4821](https://github.com/neovim/neovim/actions/runs/3812943263/jobs/6486418411#step:14:4822)
/home/runner/work/neovim/neovim//build/bin/nvim -u NONE -U NONE -i NONE --noplugin --headless -S summarize.vim messages
```

smjonas commented:
Do we really need a separate `format` variable as of now? Imo this adds unnecessary complexity. It might make sense to readd it if more formats get added in the future. To map `vim.g.tex_format` to the correct filetype, we can use a table. 

Minor suggestion: use the `format` variable directly instead of `mstr` or at least rename it to make the meaning clearer?

gi1242 commented:
Sorry to bug you guys about this so often. I'm still having trouble with your test system. Looks like the test on line 10 failed. That test creates a file called `Xfile.tex` with contents

    \newcommand{\test}{some text}
    
and then requires the filetype is `tex`. When the test runs it appears to give `plaintex` as the filetype, and I couldn't figure out why. So I created the above file, and opened it with

    install/bin/nvim -R -u NONE --cmd 'filetype on' ~/tmp/Xfile.tex -c 'echo &filetype'
    
and I got the filetype to be `tex` ...

If you know what's going on, or have other suggestions, please let me konw. I seem to have a better idea of what you expect now, so hopefully I can get a pull request in the right format once I figure out how to debug the test... 😄

gi1242 commented:
Never mind. I see the problem. `g:tex_flavor` seems to be set when I run `nvim -u NONE`, but not in the test. I'll push a fix soon (with working tests)

gi1242 commented:
Ok. All tests should pass now, and I renamed `mstr` to `format` as you requested. (I can't seem to avoid it entirely.)

Just FYI, the context detection previously there was also implemented wrong. The Vim used the 'p' search flag which made the return value the matching subexpression. This doesn't seem possible with the lua `matchstring` function, and as far as I could tell it would never match. Not to mention the starting line offset was off by one.

I think I have everything fixed now.

It also WorksForMe™, so 😁

zeertzjq commented on runtime/lua/vim/filetype/detect.lua:
```diff
@@ -1256,45 +1256,41 @@ end
-            matchregex(line, [[\c^\s*\\\%(]] .. lpat .. [[\)\|^\s*\\\(]] .. cpat .. [[\)]])
-          if lpat_match then
+        for _, line in ipairs(getlines(bufnr, i, i + 1000)) do
+          if matchregex(line, [[\c^\s*\\\%(]] .. lpat .. [[\)]] ) then
```
```suggestion
          if matchregex(line, [[\c^\s*\\\%(]] .. lpat .. [[\)]]) then
```

zeertzjq commented on runtime/lua/vim/filetype/detect.lua:
```diff
@@ -1256,45 +1256,41 @@ end
+          if matchregex(line, [[\c^\s*\\\%(]] .. lpat .. [[\)]] ) then
             return 'tex'
-          elseif cpat_match then
+          elseif matchregex(line, [[\c^\s*\\\%(]] .. cpat .. [[\)]] ) then
```
```suggestion
          elseif matchregex(line, [[\c^\s*\\\%(]] .. cpat .. [[\)]]) then
```

zeertzjq reviewed (commented):
lint failure

gi1242 commented:
Lint failure again... Looks never ending. Pushed hopefully the last fix.

smjonas commented on runtime/lua/vim/filetype/detect.lua:
```diff
@@ -1256,17 +1256,15 @@ end
     return 'context'
   else
+    -- Default value, may be changed later:
+    format = vim.g.tex_flavor and vim.g.tex_flavor or 'plaintex'
```
Can be simplified to `format = vim.g.tex_flavor or 'plainex'`

gi1242 replied:
Indeed, done (force pushed)
